### PR TITLE
Add runtime type introspection section

### PR DESCRIPTION
**Pull request by methylDragon**

### CODE REVIEW

**Comment by methylDragon** [comment]
I just saw the reviews from #1 and #2, let me push some changes

**Comment by wjwwood** [comment]
I changed the base to point to the branch for https://github.com/wjwwood/rep/pull/2, so that you don't have to keep them in sync. Once it is merged we can change the base back to the target branch and merge this one (after iterating).

**Comment by wjwwood** [comment]
@methylDragon can you rebase this, I had to rename the file due our REP number getting swiped (it's my fault).

**Comment by methylDragon** [comment]
Rebased!

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]: Technically the ability to detect type mismatches doesn't support runtime type introspection. The feature could be used to parse a message from a bagfile using just the description included with it. That neither uses a version hash nor does it require type description distribution.

It's true that we do want to use those two things in conjunction with runtime type introspection, but technically they are orthogonal. So this isn't wrong, we should just reword it a bit with better qualifying statements.

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]:
```suggestion
Specifically, this feature should allow access to the data in the fields of a serialized message buffer when given:
```

Or something to that effect, the data and fields aren't really different things, unless you mean the field names/types, which are separate, but if I say "give me the field 'x'", that generally implies the value/data.

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]:
```suggestion
- the serialized message as a ``rmw_serialized_message_t``, a.k.a. a ``rcutils_uint8_array_t``
```

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]:
```suggestion
- the message's type description, e.g. received from the aforementioned type description distribution or from a bag file
```

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]:
```suggestion
- the serialization format, name and version, which was used to create the serialized message, e.g. ``XCDR2`` for ``Extended CDR encoding version 2``
```

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]:
```suggestion
- access to the field by name or index
```

The "as a type" was strange to me, I was going to suggest "as a value" which sounds more correct to me, but then decided to suggest to remove it all together.

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]: Yeah, this would be really nice, we need to investigate more about how feasible this is. I think it's kind of necessary, since you may not have the required rmw implementations on hand all the time. It would be nice to have something completely independent that is just a library of serialization functions. But at the same time this may require us to do something like make a hard dependency on Fast-CDR (just as an example) which we otherwise would not need to do.

```suggestion
Given that the scope of inputs and expected outputs is so limited, this feature should ideally be implemented as a separate package, e.g. ``rcl_serialization``, that can be called independently by any downstream packages that might need runtime type introspection, e.g. introspection tools, rosbag transport, etc.
```

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]: I think we need to tighten up our terminology here, let's talk about it off-line.

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]: I think this couples a bit too much with transport. While most of this is true, it's also true that no transport is required. You can read a buffer and description from a bag file and access the contents of the message. No pub/sub involved. Let's rework this with that in mind.

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]:
```suggestion
As runtime type introspection is expected to work across any serialization format, the runtime type introspection interface needs to be extensible so that the necessary serialization libraries can be loaded to process the serialized message.
Serialization format support in this case will be provided by writing plugins that wrap the serialization libraries that can then provide the runtime type introspection feature with the needed facilities.
Then, runtime type introspection will:
```

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst:386]: Mention the use case for this, i.e. connext's CDR library vs Fast-CDR, which should you use if both are available?

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]: Let's try using a concrete example, like `/scan` and `sensor_msgs/msg/LaserScan`. (or image or point cloud)

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst]: Based on off-line discussion, let's drop this section.

File: rep-2011.rst
Comment (wjwwood) [inline rep-2011.rst:483]: All this needs to be updated with the nested types work.

**Review by wjwwood — APPROVED** [review]
Let's fixup what makes sense here, and then continue with the nested types pr on top.

File: rep-2011.rst
Comment (methylDragon) [reply rep-2011.rst:483]: Will be updated in the next PR

File: rep-2011.rst
Comment (methylDragon) [reply rep-2011.rst]: https://github.com/wjwwood/rep/pull/3/commits/bdfa2cfcd926334b1f28845d91b66867ab9218a6

File: rep-2011.rst
Comment (methylDragon) [reply rep-2011.rst]: https://github.com/wjwwood/rep/pull/3/commits/bdfa2cfcd926334b1f28845d91b66867ab9218a6

File: rep-2011.rst
Comment (methylDragon) [reply rep-2011.rst:386]: https://github.com/wjwwood/rep/pull/3/commits/bdfa2cfcd926334b1f28845d91b66867ab9218a6

File: rep-2011.rst
Comment (methylDragon) [reply rep-2011.rst]: https://github.com/wjwwood/rep/pull/3/commits/bdfa2cfcd926334b1f28845d91b66867ab9218a6

**Pull request merged**
